### PR TITLE
Bootstrap decomp of Tgl rendering library

### PR DESCRIPTION
**Pull request by stravant**
This PR kicks off work on decompiling the D3D Retained Mode (D3DRM) rendering part of the codebase, in the form of the `Tgl` wrapper library in the existing known source code for the game.

* I did not copy the cursed template-based template method pattern used for implementations in the original Tgl source, though I did copy the class and parameter naming.

* The base Tgl library consists of a set of abstract classes which could have multiple implementations.

* In the shipped game, there is only one implementation, using D3D Retained Mode.

* There are a few significant diversions from the known source code for Tgl. The main one is that Mesh seems to have been split into separate MeshBuilder / MeshInstance classes though we need further decomping to verify that.

* Every class contains at least one non-trivial 100% matching method call to verify things, and only a few methods have been left as STUBs. In total this PR adds 86 methods, 48 of which are 100% matches.

### CODE REVIEW

**Comment by stravant** [comment]
Didn't know about TGL. I'll do some adjustments using the TGL code to fill in some Unknowns I had here.

File: CMakeLists.txt [hunk 1]
```diff
@@ -233,6 +239,7 @@ endif()
 if (ISLE_USE_DX5)
   target_include_directories(lego1 PRIVATE "${CMAKE_SOURCE_DIR}/3rdparty/dx5/inc")
   target_link_directories(lego1 PRIVATE "${CMAKE_SOURCE_DIR}/3rdparty/dx5/lib")
+  target_link_libraries(lego1 PRIVATE d3drm)
```
Comment (madebr) [inline CMakeLists.txt:242]: Mingw needs this as well, so add it to the next `target_link_libraries` unconditionally.

Comment (stravant) [reply CMakeLists.txt]: Done.

**Comment by stravant** [comment]
Overhauled the changes to be structured like the original Tgl code.

File: LEGO1/tgl/tgl.h [hunk 2]
```diff
@@ -139,98 +123,69 @@ class Renderer : public Object {
...
+	virtual Texture* CreateTexture() = 0;
 	virtual Texture* CreateTexture(
-		int width,
-		int height,
-		int bitsPerTexel,
-		const void* pTexels,
-		int pTexelsArePersistent,
-		int paletteEntryCount,
-		const PaletteEntry* pEntries
+		int p_width,
```
Comment (foxtacles) [inline LEGO1/tgl/tgl.h:133]: For the utility libraries, the naming conventions we have in the rest of the code base don't apply:

https://github.com/isledecomp/isle/blob/master/CONTRIBUTING.md#naming-conventions

Instead, we want to retain the original naming scheme as far as it is available (even if inconsistent), so in this case we should leave all the parameter names as they were in the 1996 source. Thank you!

File: LEGO1/tgl/d3drm/group.cpp [hunk 3]
```diff
@@ -0,0 +1,129 @@
+#include "impl.h"
+
+using namespace TglImpl;
+
+// Inlined only
+GroupImpl::~GroupImpl()
+{
+	if (m_data) {
+		free(m_data);
```
Comment (foxtacles) [inline LEGO1/tgl/d3drm/group.cpp:9]: This should be `delete`

File: LEGO1/tgl/d3drm/mesh.cpp [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+DECOMP_SIZE_ASSERT(D3DRMVERTEX, 0x24);
+
+DECOMP_SIZE_ASSERT(Mesh, 0x4);
+DECOMP_SIZE_ASSERT(MeshImpl, 0x8);
+
+// Inlined only
+MeshImpl::~MeshImpl()
+{
+	if (m_data) {
+		free(m_data);
```
Comment (foxtacles) [inline LEGO1/tgl/d3drm/mesh.cpp:14]: This should be `delete`

File: LEGO1/tgl/d3drm/mesh.cpp [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+	newSomething->m_data->groupIndex = index;
+	target->ImplementationData()->SetVertices(index, 0, vcount, vertexBuffer);
+	target->ImplementationData()->SetGroupTexture(index, textureRef);
+	target->ImplementationData()->SetGroupMapping(index, mapping);
+	target->ImplementationData()->SetGroupQuality(index, quality);
+	Result result = ResultVal(target->ImplementationData()->SetGroupColor(index, color));
+
+	// Cleanup
+	if (faceBuffer)
+		free(faceBuffer);
```
Comment (foxtacles) [inline LEGO1/tgl/d3drm/mesh.cpp:120]: This should be `delete` (below too)

File: LEGO1/tgl/d3drm/texture.cpp [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+	if (p_palette != NULL) {
+		InitializePalette(p_paletteSize, p_palette);
+	}
+}
+
+// OFFSET: LEGO1 0x100a13b0
+void TglD3DRMIMAGE::Destroy()
+{
+	if (m_texelsAllocatedByClient == 0) {
+		free(m_image.buffer1);
```
Comment (foxtacles) [inline LEGO1/tgl/d3drm/texture.cpp:84]: This should be `delete` (below too)

File: LEGO1/tgl/d3drm/texture.cpp [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+}
+
+// OFFSET: LEGO1 0x100a1550
+Result TglD3DRMIMAGE::InitializePalette(int p_paletteSize, PaletteEntry* p_palette)
+{
+	// This function is a 100% match if the PaletteEntry class is copied
+	// into into the TglD3DRMIMAGE class instead of being a global struct.
+	if (m_image.palette_size != p_paletteSize) {
+		if (m_image.palette != NULL) {
+			free(m_image.palette);
```
Comment (foxtacles) [inline LEGO1/tgl/d3drm/texture.cpp:109]: This should be `delete`

File: LEGO1/viewmanager/viewlodlist.h [hunk 8]
```diff
@@ -44,7 +44,7 @@ class ViewLODList : public LODList<ViewLOD> {
 // ??? for now, until we have symbol management
 typedef const char* ROIName;
 struct ROINameComparator {
-	bool operator()(const ROIName& rName1, const ROIName& rName2) const
+	int operator()(const ROIName& rName1, const ROIName& rName2) const
```
Comment (foxtacles) [inline LEGO1/viewmanager/viewlodlist.h:47]: They did use `bool` in the 1996 source so we should leave it as is, or is there a good reason not to? (other than the compiler warning)

**Review by foxtacles — COMMENTED** [review]
Looks good to me, just a few comments. Thank you!

File: LEGO1/tgl/d3drm/renderer.cpp [hunk 9]
```diff
@@ -0,0 +1,323 @@
+#include "impl.h"
+
+using namespace TglImpl;
+
+// OFFSET: LEGO1 0x100a15e0
+Renderer* CreateRenderer()
```
Comment (foxtacles) [inline LEGO1/tgl/d3drm/renderer.cpp:6]: This function should be part of the `Tgl` namespace as declared in `tgl.h`?

File: LEGO1/tgl/d3drm/something.cpp [hunk 10]
```diff
@@ -0,0 +1,63 @@
+#include "impl.h"
+
+using namespace TglImpl;
+
+DECOMP_SIZE_ASSERT(Something, 0x4);
+DECOMP_SIZE_ASSERT(SomethingImpl, 0x8);
+
+// Inlined only
+SomethingImpl::~SomethingImpl()
```
Comment (foxtacles) [inline LEGO1/tgl/d3drm/something.cpp:9]: We generally use the term "Unknown" (or `Unk`) in the rest of the code base for unknown stuff, so we should apply it here as well

Comment (itsmattkc) [reply LEGO1/tgl/d3drm/something.cpp:9]: In general, naming anything "Something" is pretty objectively bad practice. There are hundreds if not thousands of classes we don't know yet that could all theoretically be called "Something", but how does that help us identify them? If anything, it makes it _harder_ to distinguish them than if it was just offsets.

File: LEGO1/viewmanager/viewlodlist.h (hunk 8 above)
Comment (stravant) [reply LEGO1/viewmanager/viewlodlist.h:47]: It's a compiler error for me, not just a warning, and I couldn't find a good header to include to get bool.

The original header includes some non-standard `std.h` which presumably defined bool. What do you think I should do?

Comment (foxtacles) [reply LEGO1/viewmanager/viewlodlist.h:47]: I'm assuming you are not using MSVC 4.20? It only emits a warning about the reserved keyword `bool`, but otherwise compiles fine, presumably treating it as `int`. Or are you using different compiler settings?

Either way, depending on the case, we could add a macro `COMPAT_BOOL` that expands to different things based on the compiler used, similar to `COMPAT_CONST`

File: LEGO1/tgl/d3drm/group.cpp (hunk 3 above)
Comment (stravant) [reply LEGO1/tgl/d3drm/group.cpp]: Fixed, should actually have been `Release()`

File: LEGO1/tgl/d3drm/mesh.cpp (hunk 4 above)
Comment (stravant) [reply LEGO1/tgl/d3drm/mesh.cpp:14]: I don't think that's the case.

A delete call includes a null check, but not a set to null. If this were actually a delete call then the code would either end up with two nested if statements rather than a single one, or the code setting to null outside of the if statement.

MeshImpl is the odd one out, not following the pattern the others have of directly holding a Direct3DRM handle, so it's possible the people who came in later and made that change did use malloc / free for the holder struct instead of new / delete like the initial Tgl implementers did in the leak.

File: LEGO1/tgl/d3drm/texture.cpp (hunk 6 above)
Comment (stravant) [reply LEGO1/tgl/d3drm/texture.cpp:84]: This definitely can't be a new/delete allocation because there's no null check on the frees in the destructor.

File: LEGO1/tgl/d3drm/renderer.cpp (hunk 9 above)
Comment (stravant) [reply LEGO1/tgl/d3drm/renderer.cpp]: Yup, missed `Tgl::` on there, forgot how I was doing the namespaces.

File: LEGO1/tgl/d3drm/mesh.cpp (hunk 5 above)
Comment (stravant) [reply LEGO1/tgl/d3drm/mesh.cpp]: This one could be a new/delete, changed.

File: LEGO1/viewmanager/viewlodlist.h (hunk 8 above)
Comment (stravant) [reply LEGO1/viewmanager/viewlodlist.h:47]: Ah, sorry, I didn't notice that it was only a warning since I was dealing with another actual error too at the time.

I don't think we should have warnings we can avoid but that doesn't need to be part of this PR, I'll take out the change.

File: LEGO1/tgl/d3drm/something.cpp (hunk 10 above)
Comment (stravant) [reply LEGO1/tgl/d3drm/something.cpp]: There's only one unknown class in the Tgl namespace, the rest of the codebase explicitly uses Tgl classes as `Tgl::` so I don't think there's an issue with using an ambiguous name.

Though I'm not attached to "something", changed it to `Unk` (Using `Unknown` breaks code completion because Microsoft in their infinite wisdom decided to use `Unknown` as an enum item name)

File: LEGO1/tgl/d3drm/texture.cpp (hunk 6 above)
Comment (foxtacles) [reply LEGO1/tgl/d3drm/texture.cpp:84]: I might be missing something here, but if I replace both `free`'s in this function with `delete`, I'm getting the exact the same match/disassembly.

Could you elaborate where you see the difference exactly that ascertains that it must be `free`?

File: LEGO1/tgl/d3drm/texture.cpp (hunk 7 above)
Comment (foxtacles) [reply LEGO1/tgl/d3drm/texture.cpp:109]: When substituing `free` with `delete`, the match moves from 68% to a full 100%:

image

Again, I might be missing something here, but `delete` just seems correct.

Comment (stravant) [reply LEGO1/tgl/d3drm/texture.cpp:109]: How did you make this match?? I changed the code to new/delete but it still does not for me.

File: LEGO1/tgl/d3drm/texture.cpp (hunk 6 above)
Comment (stravant) [reply LEGO1/tgl/d3drm/texture.cpp:84]: No, you're right, I was just blundering. I got myself thinking that the ifs were the C++ `delete` adding an implicit test, but the `free` also has an embedded null check, so there's no reason for C++ to be doing that.

**Comment by foxtacles** [comment]
> Though I still can't figure out how to test the scalar_deleting_destructors. Is this what you wanted?

Yeah, the way it is done for `MeshImpl` is correct. The same can/should be done for ideally every class with a virtual table there is. Sorry for the lack of documentation and guidance on this stuff, we will improve on that soon.

**Comment by stravant** [comment]
Added checks for all the inline destructors.

File: LEGO1/tgl/d3drm/camera.cpp [hunk 11]
```diff
@@ -5,14 +5,8 @@ using namespace TglImpl;
...
-// Inlined only
-CameraImpl::~CameraImpl()
-{
-	if (m_data) {
-		m_data->Release();
-		m_data = NULL;
-	}
-}
+// OFFSET: LEGO1 0x100a2560 TEMPLATE
+// TglImpl::LightImpl::`scalar deleting destructor'
```
Comment (foxtacles) [inline LEGO1/tgl/d3drm/camera.cpp:9]: Meant to be `CameraImpl`?

File: LEGO1/tgl/d3drm/view.cpp [hunk 12]
```diff
@@ -0,0 +1,350 @@
...
+	return reinterpret_cast<ViewportAppData*>(pViewport->GetAppData());
+}
+
+inline IDirect3DRMFrame* ViewportGetLightFrame(IDirect3DRMViewport* pViewport)
+{
+	return ViewportGetData(pViewport)->m_pLightFrame;
+}
+
+// OFFSET: LEGO1 0x100a23a0 TEMPLATE
+// TglImpl::LightImpl::`scalar deleting destructor'
```
[the remark this answers is not in the document]
Comment (stravant) [reply LEGO1/tgl/d3drm/view.cpp]: Aaaa.... I can't get this right. 🙃
